wc/Wc.hs

```diff
@@ -0,0 +1,154 @@
+instance Monoid (CountBy Lines) where
+    mempty = CountLines 0
+
+instance CountModeC Int64 (CountBy Lines) where
+    fromChar c = CountLines $ if c == '\n' then 1 else 0
+    getResult (CountLines x) = x
+
+-- instances for words
+--
+instance Semigroup (CountBy Words) where
+    CountWordsEmpty <> !x = x
+    !x <> CountWordsEmpty = x
+    CountWords !l !c !False <> CountWords !False !c' !r' = CountWords l (c + c' - 1) r'
+    CountWords !l !c _ <> CountWords _ !c' !r' = CountWords l (c + c') r'
+
+instance Monoid (CountBy Words) where
+    mempty = CountWordsEmpty
+
+instance CountModeC Int64 (CountBy Words) where
+    fromChar c = if isSpace c
+                               then CountWords True 0 True
+                               else CountWords False 1 False
+    getResult CountWordsEmpty = 0
+    getResult (CountWords _ n _) = n
+
+-- instances for chars
+--
+instance Semigroup (CountBy Chars) where
+    CountChars x <> CountChars y = CountChars (x + y)
+instance Monoid (CountBy Chars) where
+    mempty = CountChars 0
+
+instance CountModeC Int64 (CountBy Chars) where
+    -- TODO figure out how to do chars and bytes differently
+    -- _probably_ do it on ByteStrings, which means we have the bytes directly
+    -- and have to get chars from that. 
+    fromChar _ = CountChars 1
+    getResult (CountChars x) = x
+
+-- instances for bytes
+--
+instance Semigroup (CountBy Bytes) where
+    CountBytes x <> CountBytes y = CountBytes (x + y)
+instance Monoid (CountBy Bytes) where
+    mempty = CountBytes 0
+
+instance CountModeC Int64 (CountBy Bytes) where
+    -- TODO figure out how to do chars and bytes differently
+    -- _probably_ do it on ByteStrings, which means we have the bytes directly
+    -- and have to get chars from that. 
+    fromChar _ = CountBytes 1
+    getResult (CountBytes x) = x
+
+-- instances for maxLineLength
+--
+instance Semigroup (CountBy MaxLineLength) where
+    CountLineLengthUnbroken x <> CountLineLengthUnbroken x' =
+        CountLineLengthUnbroken (x + x')
+    CountLineLengthUnbroken x <> CountMaxLineLength l o r = 
+        let combined = l + x 
+            in CountMaxLineLength combined (max o combined) r
+    CountMaxLineLength l o r <> CountLineLengthUnbroken x =
+        let combined = r + x
+            in CountMaxLineLength l (max o combined) combined
+    CountMaxLineLength l o r <> CountMaxLineLength l' o' r' =
+        let combined = r + l'
+            in CountMaxLineLength l (max o $ max o' combined) r'
+
+instance Monoid (CountBy MaxLineLength) where
+    mempty = CountLineLengthUnbroken 0
+
+instance CountModeC Int64 (CountBy MaxLineLength) where
+    fromChar c = if c == '\n'
+                     then CountMaxLineLength 0 0 0
+                     else CountLineLengthUnbroken 1
+    getResult (CountLineLengthUnbroken n) = n
+    getResult (CountMaxLineLength _ n _) = n
+
+testStr = "abcdef\nab\nab cd\n"
+
+totals :: [[Int64]] -> [Int64]
+totals = foldl' (zipWith (+)) (repeat 0)
```

Comment:
something here is actually not quite right: for `MaxLineLength`, we want to print the max of the maxes as the `total` line, rather than the sum of the maxes. So I need to change this part. Probably by making it a list of "Some monoid" where the elements are wrapped in the newtypes from the Monoid module, namely `Sum` and `Max`. It's probably fine to extract the result back to `Int64` with `coerce` or something